rasa_sdk/forms.py

```diff
@@ -864,6 +864,9 @@ async def validate(
 
             if isinstance(validation_output, dict):
                 slots.update(validation_output)
+                # for sequential consistency, also update tracker
+                # to make changes visible to subsequent validate_{slot_name}
+                tracker.slots.update(validation_output)
```

Comment:
I think this should work on a copy of the tracker to avoid changing the original tracker for other functions

Reply:
mhm, actually we're changing the tracker directly [here](https://github.com/RasaHQ/rasa-sdk/blob/86f41e094df9565d915d5a597dce6878ba304c1f/rasa_sdk/forms.py#L701) as well - so I think it's fine to keep it like it is

Reply:
we need to get rid of this line then though: https://github.com/RasaHQ/rasa-sdk/blob/86f41e094df9565d915d5a597dce6878ba304c1f/rasa_sdk/forms.py#L704

Reply:
Could you please suggest how to best add a test for this? I could only think of making a full Rasa example which I am not sure if it is reasonable.

Did you actually mean to remove the line "tracker.add_slots(validation_events)" as it should be redundant now? - https://github.com/RasaHQ/rasa-sdk/blob/86f41e094df9565d915d5a597dce6878ba304c1f/rasa_sdk/forms.py#L699

Reply:
Hi @cmanovit I would encourage you to add tests in `[tests.test_forms`](https://github.com/RasaHQ/rasa-sdk/blob/86f41e094df9565d915d5a597dce6878ba304c1f/tests/test_forms.py) module. There are plenty of examples that you can follow to some extent, all seem to have these steps:
- create a `TestForm` class inheriting from `FormValidationAction`and initialize it as `form`
- create an instance of Tracker and an instance of `CollectingDispatcher` which can then be used as arguments to `form.run()`
- then I guess it would make sense to assert that `tracker.slots` match the expected updated slots after validation?

Also I'm not sure L699 should be removed, the `tracker.add_slots` method updates not only `tracker.slots` but also `tracker.events`. If you do remove that line, you'd need to update tracker.events in `FormValidationAction.validate` too.

Reply:
Hi @ancalita Thank you for the tip! It seems there were already tests which check the order of the returned slots and they had failed due to the change of behavior. So, I added another commit to update those tests accordingly and they are passing again.

Reply:
If this change makes sense, then I suppose it should be similarly applied to extract_custom_slots. So, I went ahead and made the change and also added a new test for it.